cmd/worker/pubsubextender/extender.go

```diff
@@ -0,0 +1,133 @@
+package pubsubextender
+
+import (
+	"context"
+	"net/url"
+	"regexp"
+	"time"
+
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/gcppubsub"
+)
+
+const (
+	defaultGracePeriod = 60 * time.Second
+	defaultDeadline    = 300 * time.Second
+)
+
+var subscriptionPathRE = regexp.MustCompile("^projects/.+/subscriptions/.+$")
+
+type driver interface {
+	// ExtendMessageDeadline sends a request to the pubsub service to change
+	// the deadline for the supplied message to the supplied deadline duration.
+	ExtendMessageDeadline(ctx context.Context, msg *pubsub.Message, deadline time.Duration) error
+
+	// GetSubscriptionDeadline is used to get the existing deadline period for
+	// the pubsub subscription. This deadline is used by default when extending
+	// the message deadline.
+	GetSubscriptionDeadline(ctx context.Context) (time.Duration, error)
+}
+
+type Extender struct {
+	driver      driver
+	Deadline    time.Duration
+	GracePeriod time.Duration
+}
+
+func New(ctx context.Context, subURL string, sub *pubsub.Subscription) (*Extender, error) {
+	u, err := url.Parse(subURL)
+	if err != nil {
+		return nil, err
+	}
+
+	var d driver
+	switch u.Scheme {
+	case gcppubsub.Scheme:
+		d, err = newGCPDriver(u, sub)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		d = &noopDriver{}
+	}
+
+	deadline, err := d.GetSubscriptionDeadline(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if deadline == 0 {
+		deadline = defaultDeadline
+	}
+
+	return &Extender{
+		driver:      d,
+		Deadline:    deadline,
+		GracePeriod: defaultGracePeriod,
+	}, nil
+}
+
+type MessageExtender struct {
+	ticker   *time.Ticker
+	msg      *pubsub.Message
+	done     chan bool
+	exited   chan error
+	callback func()
+	running  bool
+}
+
+func (e *Extender) Start(ctx context.Context, msg *pubsub.Message, callback func()) *MessageExtender {
+	freq := e.Deadline - e.GracePeriod
+	if freq < 0 {
```

Comment:
not sure if this is the most intuitive behaviour, perhaps it should just return an error?